go/cmd/vtctldclient/command/vreplication/vdiff/vdiff_env_test.go

```diff
@@ -0,0 +1,355 @@
+	})
+	tabletconntest.SetProtocol("go.cmd.vtctldclient.vreplication.vdiff_env_test", dialerName)
+
+	tabletID := 100
+	for _, shard := range sourceShards {
+		_ = env.addTablet(tabletID, env.sourceKeyspace, shard, topodatapb.TabletType_PRIMARY)
+		env.tmc.waitpos[tabletID+1] = vdiffStopPosition
+
+		tabletID += 10
+	}
+	tabletID = 200
+	for _, shard := range targetShards {
+		primary := env.addTablet(tabletID, env.targetKeyspace, shard, topodatapb.TabletType_PRIMARY)
+
+		var rows []string
+		var posRows []string
+		for j, sourceShard := range sourceShards {
+			bls := &binlogdatapb.BinlogSource{
+				Keyspace: env.sourceKeyspace,
+				Shard:    sourceShard,
+				Filter: &binlogdatapb.Filter{
+					Rules: []*binlogdatapb.Rule{{
+						Match:  "t1",
+						Filter: query,
+					}},
+				},
+			}
+			rows = append(rows, fmt.Sprintf("%d|%v|||", j+1, bls))
+			position := vdiffStopPosition
+			if pos := positions[sourceShard+shard]; pos != "" {
+				position = pos
+			}
+			posRows = append(posRows, fmt.Sprintf("%v|%s", bls, position))
+
+			// vdiff.syncTargets. This actually happens after stopTargets.
+			// But this is one statement per stream.
+			env.tmc.setVRResults(
+				primary.tablet,
+				fmt.Sprintf("update _vt.vreplication set state='Running', stop_pos='%s', message='synchronizing for vdiff' where id=%d", vdiffSourceGtid, j+1),
+				&sqltypes.Result{},
+			)
+		}
+		// migrater buildMigrationTargets
+		env.tmc.setVRResults(
+			primary.tablet,
+			"select id, source, message, cell, tablet_types, workflow_type, workflow_sub_type, defer_secondary_keys from _vt.vreplication where workflow='vdiffTest' and db_name='vt_target'",
+			sqltypes.MakeTestResult(sqltypes.MakeTestFields(
+				"id|source|message|cell|tablet_types|workflow_type|workflow_sub_type|defer_secondary_keys",
+				"int64|varchar|varchar|varchar|varchar|int64|int64|int64"),
+				rows...,
+			),
+		)
+
+		// vdiff.stopTargets
+		env.tmc.setVRResults(primary.tablet, "update _vt.vreplication set state='Stopped', message='for vdiff' where db_name='vt_target' and workflow='vdiffTest'", &sqltypes.Result{})
+		env.tmc.setVRResults(
+			primary.tablet,
+			"select source, pos from _vt.vreplication where db_name='vt_target' and workflow='vdiffTest'",
+			sqltypes.MakeTestResult(sqltypes.MakeTestFields(
+				"source|pos",
+				"varchar|varchar"),
+				posRows...,
+			),
+		)
+
+		// vdiff.syncTargets (continued)
+		env.tmc.vrpos[tabletID] = vdiffSourceGtid
+		env.tmc.pos[tabletID] = vdiffTargetPrimaryPosition
+
+		// vdiff.startQueryStreams
+		env.tmc.waitpos[tabletID+1] = vdiffTargetPrimaryPosition
+
+		// vdiff.restartTargets
+		env.tmc.setVRResults(primary.tablet, "update _vt.vreplication set state='Running', message='', stop_pos='' where db_name='vt_target' and workflow='vdiffTest'", &sqltypes.Result{})
+
+		tabletID += 10
+	}
+	env.resetOutput()
+	return env
+}
+
+func (env *testVDiffEnv) resetOutput() {
+	env.mu.Lock()
+	defer env.mu.Unlock()
+	ogstdout := os.Stdout
```

Comment:
ngl i love this name a lot

Reply:
that said, i think at least the tests for `displaySimple<blah blah i forget the full name>` could be simplified by having it return a string rather than printing, then updating the callsites to `fmt.Println(SimpleReport())`

Reply:
The real reason for the unit tests is the `show` output, which is complex and the real critical part on the client side. For that we print lines as we go along. So the unit tests would remain the same — although I do like printing something in the command itself too.

I also like copying STDOUT because that's exactly what the user sees as the results, there's no ambiguity or chance for the test to misalign with the user experience.

Reply:
Just noting that you convinced me of a better alternative on Slack so I will work on that 🙂 . Thanks!

Reply:
I took your good suggestion and added a writer: https://github.com/vitessio/vitess/pull/13976/commits/3acb0c126cd7cea480c5875dfd4299aea918e7e4

Thanks again!